### PR TITLE
Pr 22449

### PR DESCRIPTION
**Pull request by mhvk**
@seberg - see main PR for what the intent is here.

### CODE REVIEW

**Comment by seberg** [comment]
Thanks, sorry its a bit messy :(.  I have to look closer at it, but don't mind removing fmt.  Part of the reason for why I put it is that I thought it may be nicer to let the dtypes deal with it for actual format strings.

I.e. there was/is a PR to add support for `f"{arr:4.3f}"` and others, and at the time, it seemed messy to me to implement it just by updating `printoptions`.  For one, a float format doesn't make sense for strings or datetimes!
So, on top of just adding the quotes, the idea was to go in a direction where we can support format strings clearer in a way that can tell the dtype to decide what the format option means!

**Comment by mhvk** [comment]
Agreed that a nicer way to format user dtypes would help, but I think it is better left to a separate PR. (A more minimal improvement that I would like is to make structured and subarrays controllable as well. But perhaps the easiest way to do that is to allow dtypes as entries in the `formatter` dict.)

**Comment by seberg** [comment]
OK, I like the simplification.  After reviewing, I realized the initial reason for why I added this annoyance:
```
arr = np.array([1], dtype=np.longdouble)
print(repr(arr))
print(str(arr))
```
I wanted quotes for `repr` but not `str`.  And to do that, I need to pass in `repr=True/False` somehow (which I did, maybe oddly overloaded via `fmt=`).

OTOH, I now also realized that is a bit dubious because the quotes because the default precision when printing arrays is low, and the quotes are misleading unless we print with full precision. 

We are losing the quotes on `np.void(3, dtype="i,g")`, although that should be easy enough to recover.

---

Maybe we should just could cut our losses in the first iteration and add the quotes only to the scalar `repr` but not to the longdouble "item repr", considering that longdouble is an odd-one-out anyway?

**Comment by mhvk** [comment]
@seberg - agreed, for a longdouble array repr, quoting makes less sense, since there never is the idea that the repr will round-trip (unlike for the scalar). So, probably should be user-controllable if we do it at all.

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -402,39 +399,38 @@ def _object_format(o):
...
 def repr_format(x):
-    if isinstance(x, (np.str_, np.bytes_)):
+    if isinstance(x, (np.str_, np.bytes_)) and _format_options['legacy'] > 125:
         return repr(x.item())
     return repr(x)
 
 def str_format(x):
-    if isinstance(x, (np.str_, np.bytes_)):
+    if isinstance(x, (np.str_, np.bytes_)) and _format_options['legacy'] > 125:
         return str(x.item())
```
Comment (seberg) [inline numpy/core/arrayprint.py:408]: Not tha tit matters, but I think this is unnecessary (and thus easier to remove)?!  (`item()` works for the old version, too).

**Comment by seberg** [comment]
OK, so for `void`, I would also just return the wrong thing for now (no quotes).  I still suspect adding `fmt` is probably the right way (as much as I dislike it), but I am mor than happy to defer it, it makes things annoying.

(I also think this should be a proper contextvar and that once it is, maybe we should just stop passing `options` around at all, after which a new `fmt` might seem less weird).

Sounds fine?  (in which case I will merge this and make an iteration)

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (mhvk) [reply numpy/core/arrayprint.py:408]: Sure, if the output is identical, there is no reason to have it!

**Comment by mhvk** [comment]
I think it is indeed best to just merge this now. For `void`, if it is easy, maybe it is nice o make it round-trip `longdouble` too, though in the end that can be lower priority, since we know `void` won't round-trip if there is a sub-array in it anyway.

File: numpy/ma/core.py [hunk 2]
```diff
@@ -4076,11 +4076,10 @@ def __repr__(self):
             # The dtype of the fill value should match that of the array
             # and it is unnecessary to print the full `np.int64(...)`
             # Allow strings: "N/A" has length 3 so would mismatch often.
-            fill_repr = np.core.arrayprint.get_formatter(
-                    dtype=self._fill_value.dtype, fmt=repr)(self.fill_value)
+            fill_repr = str(self.fill_value)
```
Comment (seberg) [inline numpy/ma/core.py:4079]: This doesn't work for strings, we need to do the same thing we do inside structured voids here really, but that is hard now, so I will add a branch for strings and live with the rest of the fallout.

**Pull request merged**